### PR TITLE
Change collection PS util import pattern

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
Change the way PowerShell and C# module utils inside a collection are referenced. This is a breaking change but is being done to unify import patterns between the Python, PowerShell, and C#. Collections are a tech preview in 2.8 so this should hopefully be in by 2.9 before the existing pattern is more solidified.

This PR also adds a way to specify a type accelerator inside a C# util to simplify the type name a bit more. This is optional and a suggestion to add to Ansible but can have consequences if 2 utils use the same accelerator.

This also requires some changes in `ansible-test` but waiting to see what @nitzmahone and @mattclay says before delving in there.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
PowerShell
Collections

### CODE REVIEW

File: lib/ansible/executor/powershell/module_manifest.py [hunk 1]
```diff
@@ -35,10 +37,31 @@ def __init__(self):
         self.os_version = None
         self.become = False
 
-        self._re_cs_module = re.compile(to_bytes(r'(?i)^using\s((Ansible\..+)|(AnsibleCollections\.\w.+\.\w.+\w.+));\s*$'))
-        self._re_cs_in_ps_module = re.compile(to_bytes(r'(?i)^#\s*ansiblerequires\s+-csharputil\s+((Ansible\..+)|(AnsibleCollections\.\w.+\.\w.+\w.+))'))
-        self._re_coll_ps_in_ps_module = re.compile(to_bytes(r'(?i)^#\s*ansiblerequires\s+-powershell\s+((Ansible\..+)|(AnsibleCollections\.\w.+\.\w.+\w.+))'))
-        self._re_module = re.compile(to_bytes(r'(?i)^#\s*requires\s+\-module(?:s?)\s*(Ansible\.ModuleUtils\..+)'))
+        self._re_cs_module = [
+            # Reference C# module_util in another C# util
+            re.compile(to_bytes(r'(?i)^using\s((Ansible\..+)|'
```
Comment (mattclay) [inline lib/ansible/executor/powershell/module_manifest.py:48]: Include an example in the comment like the comments for the patterns below provide.

File: docs/docsite/rst/porting_guides/porting_guide_2.9.rst [hunk 2]
```diff
@@ -34,6 +34,29 @@ Deprecated
 No notable changes
 
 
+Collection loader changes
+=========================
+
+The way to import a PowerShell of C# module util from a collection has changed in the Ansible 2.9 release. In Ansible
```
Comment (mattclay) [inline docs/docsite/rst/porting_guides/porting_guide_2.9.rst]:
```suggestion
The way to import a PowerShell or C# module util from a collection has changed in the Ansible 2.9 release. In Ansible
```

File: docs/docsite/rst/dev_guide/collections_tech_preview.rst [hunk 3]
```diff
@@ -86,10 +86,10 @@ module_utils
 
 When coding with ``module_utils`` in a collection, the Python ``import`` statement needs to take into account the FQCN along with the ``ansible_collections`` convention. The resulting Python import will look like ``from ansible_collections.{namespace}.{collection}.plugins.module_utils.{util} import {something}``
 
-The following example snippet shows a module using both default Ansible ``module_utils`` and
-those provided by a collection. In this example the namespace is
-``ansible_example``, the collection is ``community``, and the ``module_util`` in
-question is called ``qradar`` such that the FQCN is ``ansible_example.community.plugins.module_utils.qradar``:
+The following example snippets shows a Python and PowerShell module using both default Ansible ``module_utils`` and
```
Comment (mattclay) [inline docs/docsite/rst/dev_guide/collections_tech_preview.rst]:
```suggestion
The following example snippets show a Python and PowerShell module using both default Ansible ``module_utils`` and
```

**Review by mattclay — COMMENTED** [review]
Looks good overall. A few minor changes to comments requested.

As for the type accelerators feature, that should probably be in a separate PR so it can be handled independently of this change.

File: lib/ansible/executor/powershell/module_manifest.py [hunk 4]
```diff
@@ -35,10 +37,31 @@ def __init__(self):
...
+                                r'(ansible_collections\.[\w.]+\.[\w.]+\.plugins\.module_utils\.[\w.]+));\s*$')),
```
Comment (nitzmahone) [inline lib/ansible/executor/powershell/module_manifest.py]: Why include the `.` in the character groups? `[\w.]` is the same as `[.]`- and allows lots of things we don't want...

File: lib/ansible/executor/powershell/module_manifest.py [hunk 5]
```diff
@@ -35,10 +37,31 @@ def __init__(self):
...
+        ]
+
+        self._re_cs_in_ps_module = [
+            # Reference C# module_util in a PowerShell module
+            # '#AnsibleRequires -CSharpUtil Ansible.{name}'
+            # '#AnsibleRequires -CSharpUtil ansible_collections.{namespace}.{collection}.plugins.module_utils.{name}'
+            re.compile(to_bytes(r'(?i)^#\s*ansiblerequires\s+-csharputil\s+((Ansible\..+)|'
```
Comment (nitzmahone) [inline lib/ansible/executor/powershell/module_manifest.py:56]: ditto on `.` in character groups

File: lib/ansible/executor/powershell/module_manifest.py [hunk 6]
```diff
@@ -35,10 +37,31 @@ def __init__(self):
...
+
+        self._re_ps_module = [
+            # Original way of referencing a builtin module_util
+            # '#Requires -Module Ansible.ModuleUtils.{name}
+            re.compile(to_bytes(r'(?i)^#\s*requires\s+\-module(?:s?)\s*(Ansible\.ModuleUtils\..+)')),
+            # New way of referencing a builtin and collection module_util
+            # '#AnsibleRequires -PowerShell ansible_collections.{namespace}.{collection}.plugins.module_utils.{name}'
+            # '#AnsibleRequires -PowerShell Ansible.ModuleUtils.{name}'
+            re.compile(to_bytes(r'(?i)^#\s*ansiblerequires\s+-powershell\s+(((Ansible\.ModuleUtils\..+))|'
+                                r'(ansible_collections\.[\w.]+\.[\w.]+\.plugins\.module_utils\.[\w.]+))')),
```
Comment (nitzmahone) [inline lib/ansible/executor/powershell/module_manifest.py]: ditto on `.` in character groups

**Review by nitzmahone — COMMENTED** [review]
Looks good- one thing to consider:

As a developer quality-of-life thing, it might be good to split the "find the #AnsibleRequires lines" away from the "load the contents" (not necessarily for freeze either, but as a bugfix/enhancement after), so that a malformed requires statement can give an error instead of being silently ignored because the regex didn't find it. It'd also probably make the regexes a little more "bite-size" and easier to follow.

I definitely like the "less magic" approach here to what I originally did, and the type accelerator thing is a good way to make that feature more accessible to stop the madness on verbosity of static/type refs.

File: lib/ansible/executor/powershell/module_manifest.py (hunk 4 above)
Comment (jborean93) [reply lib/ansible/executor/powershell/module_manifest.py]: The original regex originally was `\w.+\.` which mean match 1 word char then allow any char until `.`. Happy to change this to be just `\w+` to match `[A-Za-z0-0_]` but having the `.` in there before made me pause.

**Comment by jborean93** [comment]
> As a developer quality-of-life thing, it might be good to split the "find the #AnsibleRequires lines" away from the "load the contents" (not necessarily for freeze either, but as a bugfix/enhancement after)

Sounds fair enough, I'll add a comment for future work.

> I definitely like the "less magic" approach here to what I originally did, and the type accelerator thing is a good way to make that feature more accessible to stop the madness on verbosity of static/type refs.

I thought it was a good idea and potentially a nice way to mask a potentially namespace/class change of the builtin C# utils if we ever do that. My only misgiving is that it potentially allows someone to have a name conflict if they don't really use something unique.

**Review by mattclay — APPROVED** [review]

**Review by nitzmahone — APPROVED** [review]

**Pull request merged**